packages/tpetra/core/example/advanced/Benchmarks/CrsMatrixMatVec.cpp

```diff
@@ -0,0 +1,380 @@
+  // Contrary to expectations, asking for the graph's number of
+  // columns, or asking the column Map for the number of entries,
+  // won't give the correct number of columns in the graph.
+  // const GO gblNumCols = graph->getDomainMap ()->getGlobalNumElements ();
+  const LO lclNumRows = meshRowMap.getNodeNumElements ();
+
+  RCP<matrix_type> A = rcp (new matrix_type (graph));
+
+  // Fill in the sparse matrix.
+  out << "Fill the CrsMatrix" << endl;
+  for (LO lclRow = 0; lclRow < lclNumRows; ++lclRow) { // for each of my rows
+    Teuchos::ArrayView<const LO> lclColInds;
+    graph->getLocalRowView (lclRow, lclColInds);
+
+    // Put some entries in the matrix.
+    Teuchos::Array<SC> lclValues(lclColInds.size(), Teuchos::ScalarTraits<SC>::one());
+    const LO err = A->replaceLocalValues (lclRow, lclColInds, lclValues());
+    TEUCHOS_TEST_FOR_EXCEPTION(err != lclColInds.size(), std::logic_error, "Bug");
+  }
+  A->fillComplete();
+
+  return A;
+}
+
+} // namespace (anonymous)
+
+
+int
+main (int argc, char* argv[])
+{
+  using Teuchos::RCP;
+  using Teuchos::REDUCE_MIN;
+  using Teuchos::reduceAll;
+  using Teuchos::outArg;
+  using Teuchos::TimeMonitor;
+  using std::endl;
+  typedef Tpetra::Vector<>::scalar_type SC;
+
+  Tpetra::ScopeGuard tpetraScope (&argc, &argv);
+  bool success = true;
+  {
+    auto comm = Tpetra::getDefaultComm ();
+
+    // Output stream 'out' will ignore output not from Process 0.
+    RCP<Teuchos::FancyOStream> pOut = getOutputStream (*comm);
+    Teuchos::FancyOStream& out = *pOut;
+
+    // Read command-line options into the 'opts' struct.
+    CmdLineOpts opts;
+    {
+      Teuchos::CommandLineProcessor clp;
+      setCmdLineOpts (opts, clp);
+      int result = parseCmdLineOpts (clp, argc, argv);
+      if (result == 1) { // help printed
+        return EXIT_SUCCESS;
+      }
+      else if (result == -1) { // parse error
+        return EXIT_FAILURE;
+      }
+      result = checkCmdLineOpts (out, *comm, opts);
+      if (result != 0) {
+        return EXIT_FAILURE;
+      }
+    }
+
+    out << "Command-line options:" << endl;
+    printCmdLineOpts (out, opts);
+
+    auto G = getTpetraGraph (comm, opts);
+    auto A = getTpetraCrsMatrix (out, G, opts);
+    Tpetra::Vector<> X (A->getDomainMap ());
+    Tpetra::Vector<> Y (A->getRangeMap ());
+
+    // Fill X with values that don't increase the max-norm of results.
+    // That way, repeated mat-vecs won't overflow.  This matters
+    // because some processors do a silly thing and handle Inf or NaN
+    // (or even denorms) via traps.  This is very expensive, so if the
+    // norms increase or decrease a lot, that might trigger the slow
+    // case.
+    const SC X_val = static_cast<SC> (1.0) /
+      static_cast<SC> (opts.numEntPerRow);
+    X.putScalar (X_val);
+    Y.putScalar (0.0);
+
+    auto timer = TimeMonitor::getNewCounter ("Tpetra CrsMatrix apply (mat-vec)");
+    {
+      TimeMonitor timeMon (*timer);
```

Comment:
We should do an apply to "warm up" the UVM allocations, since the code fills the matrix on host.

Reply:
Yes, that's a good idea, I'll add that.